### PR TITLE
Per Activity Method Options (Part I)

### PR DESCRIPTION
**Pull request by vkoby**
Added ability to pass activity method options to activityStub. Usage:

Map<activityTypeName, ActivityOptions> activityMethodOptions
Map<activityTypeName, LocalActivityOptions> localActivityMethodOptions

Workflow.newActivityStub(Activity.class, options, activityMethodOptions)
Workflow.newLocalActivityStub(Activity.class, options, localActivityMethodOptions)

WIP: Ability to pass activity method options with WorkflowImplementationOptions. 

Fix for issue #370

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java [hunk 1]
```diff
@@ -54,8 +60,13 @@ private ActivityInvocationHandler(
...
       Method method, MethodRetry methodRetry, String activityName) {
     Function<Object[], Object> function;
+    ActivityOptions methodOptions =
+        (activityMethodOptions == null) ? null : activityMethodOptions.get(method.getName());
     ActivityOptions mergedOptions =
-        ActivityOptions.newBuilder(options).mergeMethodRetry(methodRetry).build();
+        ActivityOptions.newBuilder(options)
+            .mergeMethodOptions(methodOptions)
+            .mergeMethodRetry(methodRetry)
+            .build();
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java:69]: Move merging to the constructor.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java [hunk 2]
```diff
@@ -54,8 +60,13 @@ private ActivityInvocationHandler(
   protected Function<Object[], Object> getActivityFunc(
       Method method, MethodRetry methodRetry, String activityName) {
     Function<Object[], Object> function;
+    ActivityOptions methodOptions =
+        (activityMethodOptions == null) ? null : activityMethodOptions.get(method.getName());
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java:64]: I think we want to use the activity type name as the key to the map, not the method name.

File: temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java [hunk 3]
```diff
@@ -379,16 +388,21 @@
...
+   * @param activityInterface interface type implemented by activities.
+   * @param options options that together with the properties of {@link
+   *     io.temporal.activity.ActivityMethod} specify the activity invocation parameters.
    */
-  public static <T> T newActivityStub(Class<T> activityInterface) {
-    return WorkflowInternal.newActivityStub(activityInterface, null);
+  public static <T> T newActivityStub(
+      Class<T> activityInterface,
+      ActivityOptions options,
+      Map<String, ActivityOptions> perActivityMethodOptions) {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:404]: Rename to perActivityTypeActivityOptions

File: temporal-sdk/src/main/java/io/temporal/activity/ActivityOptions.java [hunk 4]
```diff
@@ -175,6 +175,44 @@ public Builder setCancellationType(ActivityCancellationType cancellationType) {
       return this;
     }
 
+    public Builder mergeMethodOptions(ActivityOptions methodOptions) {
```
[the remark this answers is not in the document]
Comment (vkoby) [reply temporal-sdk/src/main/java/io/temporal/activity/ActivityOptions.java]: There's a mergeMethodRetry below, do you think we should follow the same naming pattern?

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java [hunk 5]
```diff
@@ -26,26 +26,32 @@
...
 public class ActivityInvocationHandler extends ActivityInvocationHandlerBase {
   private final ActivityOptions options;
+  private final Map<String, ActivityOptions> activityMethodOptions;
   private final WorkflowOutboundCallsInterceptor activityExecutor;
 
   @VisibleForTesting
   public static InvocationHandler newInstance(
       Class<?> activityInterface,
       ActivityOptions options,
+      Map<String, ActivityOptions> methodOptions,
```
[the remark this answers is not in the document]
Comment (vkoby) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java]: Not sure what you meant here. Annotations?

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java (hunk 2 above)
Comment (vkoby) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java]: I remember you saying that keying from the actual Method object would be best and it's very conveniently here. Should we do that instead?

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java]: You can use @nonnull annotation. But it is not enough. I would add

this.methodOptions = Objects.requireNonNull(methodOptions);

Then you don't need this field nullness anywhere in the code that uses it.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java (hunk 2 above)
Comment (mfateev) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/ActivityInvocationHandler.java]: It is not as we planned to support the configuration of these options through a config file. In this case activity type name makes much more sense than the method name.

File: temporal-sdk/src/main/java/io/temporal/activity/ActivityOptions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply temporal-sdk/src/main/java/io/temporal/activity/ActivityOptions.java]: I see. Then I would call it `mergeActivityOptions` to follow that pattern.

File: temporal-sdk/src/main/java/io/temporal/worker/WorkflowImplementationOptions.java [hunk 6]
```diff
@@ -40,6 +42,8 @@ public static Builder newBuilder() {
   public static final class Builder {
 
     private Class<? extends Throwable>[] failWorkflowExceptionTypes;
+    private Map<String, ActivityOptions> activityOptions;
+    private Map<String, Map<String, ActivityOptions>> activityToMethodOptions;
```
[the remark this answers is not in the document]
Comment (vkoby) [reply temporal-sdk/src/main/java/io/temporal/worker/WorkflowImplementationOptions.java:46]: Yup that's for the Worker implementation

File: temporal-sdk/src/test/java/io/temporal/activity/ActivityMethodOptionsTest.java [hunk 7]
```diff
@@ -20,98 +20,135 @@
...
+        defaultOps.getStartToCloseTimeout(),
+        Duration.parse(method2OpsValues.get("StartToCloseTimeout")));
   }
 
   @ActivityInterface
   public interface TestActivity {
 
     @ActivityMethod
-    void method1();
+    Map<String, String> method1();
```
Comment (mfateev) [inline temporal-sdk/src/test/java/io/temporal/activity/ActivityMethodOptionsTest.java:118]: Change the activity methods to return Map<String, Duration> to remove toString and parse.

File: temporal-testing/src/main/java/io/temporal/testing/TestActivityEnvironment.java [hunk 8]
```diff
@@ -78,6 +81,36 @@ static TestActivityEnvironment newInstance(TestEnvironmentOptions options) {
...
   <T> T newActivityStub(Class<T> activityInterface);
 
+  /**
+   * Creates a stub that can be used to invoke activities registered through {@link
+   * #registerActivitiesImplementations(Object...)}.
+   *
+   * @param <T> Type of the activity interface.
+   * @param activityInterface activity interface class that the object under test implements
+   * @param options options that specify the activity invocation parameters
+   * @param activityMethodOptions activity method-specific invocation parameters
```
Comment (mfateev) [inline temporal-testing/src/main/java/io/temporal/testing/TestActivityEnvironment.java:91]: Document that the key value is the activity type which is by default is the method name with the first capitalized letter.

File: temporal-testing/src/main/java/io/temporal/testing/TestActivityEnvironment.java [hunk 9]
```diff
@@ -78,6 +81,36 @@ static TestActivityEnvironment newInstance(TestEnvironmentOptions options) {
...
+      Map<String, ActivityOptions> activityMethodOptions);
+
+  /**
+   * Creates a stub that can be used to invoke activities registered through {@link
+   * #registerActivitiesImplementations(Object...)}.
+   *
+   * @param <T> Type of the activity interface.
+   * @param activityInterface activity interface class that the object under test implements
+   * @param options options that specify the activity invocation parameters
+   * @param activityMethodOptions activity method-specific invocation parameters
```
Comment (mfateev) [inline temporal-testing/src/main/java/io/temporal/testing/TestActivityEnvironment.java:106]: ditto

**Review by mfateev — APPROVED** [review]

**Pull request merged**